File: Apps/Database/Domain/Apps/Product/NonSerialisedInventoryItem.cs

// <copyright file="NonSerialisedInventoryItem.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Database.Domain
{
    public partial class NonSerialisedInventoryItem
    {
        // TODO: Cache
        public TransitionalConfiguration[] TransitionalConfigurations => new[] {
            new TransitionalConfiguration(this.M.NonSerialisedInventoryItem, this.M.NonSerialisedInventoryItem.NonSerialisedInventoryItemState),
        };

        public void AppsOnBuild(ObjectOnBuild method)
        {
            if (!this.ExistNonSerialisedInventoryItemState)
            {
                this.NonSerialisedInventoryItemState = new NonSerialisedInventoryItemStates(this.Strategy.Transaction).Good;
            }
        }

        public void AppsDelete(DeletableDelete method)
        {
            foreach (InventoryItemVersion version in this.AllVersions)
            {
                version.Delete();
            }
        }

        public decimal CalculateQuantityOnHand(Settings settings)
        {
            var quantityOnHand = 0M;

            if (!settings.InventoryStrategy.OnHandNonSerialisedStates.Contains(this.NonSerialisedInventoryItemState))
            {
                this.QuantityOnHand = 0;
            }
            else
            {
                foreach (InventoryItemTransaction inventoryTransaction in this.InventoryItemTransactionsWhereInventoryItem)
                {
                    var reason = inventoryTransaction.Reason;

                    if (reason?.IncreasesQuantityOnHand == true)
                    {
                        quantityOnHand += inventoryTransaction.Quantity;
                    }
                    else if (reason?.IncreasesQuantityOnHand == false)
                    {
                        quantityOnHand -= inventoryTransaction.Quantity;
                    }
                }

                foreach (PickListItem pickListItem in this.PickListItemsWhereInventoryItem)
                {
                    if (pickListItem.PickListWherePickListItem.PickListState.Equals(new PickListStates(this.Strategy.Transaction).Picked))
                    {
                        foreach (ItemIssuance itemIssuance in pickListItem.ItemIssuancesWherePickListItem)
                        {
                            if (!itemIssuance.ShipmentItem.ShipmentItemState.IsShipped)
                            {
                                quantityOnHand -= pickListItem.QuantityPicked;
                            }
                        }
                    }
                }
            }

            return quantityOnHand;
        }

        public decimal CalculateQuantityCommittedOut()
        {
            var quantityCommittedOut = 0M;

            foreach (InventoryItemTransaction inventoryItemTransaction in this.InventoryItemTransactionsWhereInventoryItem)
            {
                var reason = inventoryItemTransaction.Reason;

                if (reason?.IncreasesQuantityCommittedOut == true)
                {
                    quantityCommittedOut += inventoryItemTransaction.Quantity;
                }
                else if (reason?.IncreasesQuantityCommittedOut == false)
                {
                    quantityCommittedOut -= inventoryItemTransaction.Quantity;
                }
            }

            foreach (PickListItem pickListItem in this.PickListItemsWhereInventoryItem)
            {
                if (pickListItem.PickListWherePickListItem.PickListState.Equals(new PickListStates(this.Strategy.Transaction).Picked))
                {
                    foreach (ItemIssuance itemIssuance in pickListItem.ItemIssuancesWherePickListItem)
                    {
                        if (!itemIssuance.ShipmentItem.ShipmentItemState.IsShipped)
                        {
                            quantityCommittedOut -= pickListItem.QuantityPicked;
                        }
                    }
                }
            }

            if (quantityCommittedOut < 0)
            {
                quantityCommittedOut = 0;
            }

            return quantityCommittedOut;
        }

        public decimal CalculateAvailableToPromise(Settings settings)
        {
            var quantityOnHand = this.CalculateQuantityOnHand(settings);
            var quantityCommittedOut = this.CalculateQuantityCommittedOut();

            var availableToPromise = quantityOnHand - quantityCommittedOut;

            if (availableToPromise < 0)
            {
                availableToPromise = 0;
            }

            return availableToPromise;
        }
    }
}

File: _dotnet/Apps/workspace/typescript/libs/domain/apps/src/nonserialisedinventoryitem.ts
import { Meta } from '@allors/meta/generated';
import { assert } from '@allors/meta/core';
import { NonSerialisedInventoryItem } from '@allors/domain/generated';
import { Database } from '@allors/workspace/core';

export function extendNonSerialisedInventoryItem(database: Database) {

  const m = database.metaPopulation as Meta;
  const cls = database.constructorByObjectType.get(m.NonSerialisedInventoryItem);
  assert(cls);

  Object.defineProperty(cls.prototype, 'facilityName', {
    configurable: true,
    get(this: NonSerialisedInventoryItem): string  {
      return this.Facility?.Name ?? '';
    },
  });

  // Object.defineProperty(cls.prototype, 'quantityOnHand', {
  //   configurable: true,
  //   get(this: NonSerialisedInventoryItem) {
  //     let quantity = 0;

  //     this.InventoryItemTransactionsWhereInventoryItem.forEach((inventoryTransaction: InventoryItemTransaction) => {
  //       const reason = inventoryTransaction.Reason;

  //       if (reason?.IncreasesQuantityOnHand && inventoryTransaction.Quantity) {
  //         if (reason.IncreasesQuantityOnHand) {
  //           quantity += parseFloat(inventoryTransaction.Quantity);
  //         } else {
  //           quantity -= parseFloat(inventoryTransaction.Quantity);
  //         }
  //       }

  //     });

  //     return quantity;
  //   },
  // });

  // Object.defineProperty(cls.prototype, 'quantityCommittedOut', {
  //   configurable: true,
  //   get(this: NonSerialisedInventoryItem) {
  //     let quantity = 0;

  //     this.InventoryItemTransactionsWhereInventoryItem.forEach((inventoryTransaction: InventoryItemTransaction) => {
  //       const reason = inventoryTransaction.Reason;

  //       if (reason?.IncreasesQuantityCommittedOut && inventoryTransaction.Quantity) {
  //         if (reason.IncreasesQuantityCommittedOut) {
  //           quantity += parseFloat(inventoryTransaction.Quantity);
  //         } else {
  //           quantity -= parseFloat(inventoryTransaction.Quantity);
  //         }
  //       }
  //     });

  //     return quantity;
  //   },
  // });

  // Object.defineProperty(cls.prototype, 'availableToPromise', {
  //   configurable: true,
  //   get(this: NonSerialisedInventoryItem) {

  //     let quantity = this.quantityOnHand - this.quantityCommittedOut;

  //     if (quantity < 0) {
  //       quantity = 0;
  //     }

  //     return quantity;
  //   },
  // });

};
